File: app/src/main/java/com/capitalflow/app/data/Calculations.kt
package com.capitalflow.app.data

import com.capitalflow.app.model.*
import java.text.SimpleDateFormat
import java.util.Locale
import kotlin.math.max
import kotlin.math.min

object Calculations {

    private val dateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.US)

    /**
     * Computes running spreadsheet values for an investment.
     */
    fun computeMonthlyRecords(
        investment: Investment,
        records: List<MonthlyRecord>
    ): List<ComputedMonthlyRecord> {
        val sorted = records.sortedBy { rec ->
            try {
                dateFormat.parse(rec.date)?.time ?: 0L
            } catch (_: Exception) {
                0L
            }
        }

        var cumulativeCapital = investment.initialCapital
        var cumulativeRecovered = 0.0

        return sorted.map { rec ->
            val netProfit = rec.income - rec.expense
            cumulativeCapital += rec.additionalCapital - rec.withdrawal
            cumulativeRecovered += rec.income

            val remainingCapitalAfter = max(0.0, cumulativeCapital - cumulativeRecovered)
            val monthlyRoi = if (cumulativeCapital > 0.0) (netProfit / cumulativeCapital) * 100.0 else 0.0

            ComputedMonthlyRecord(
                record = rec,
                netProfit = netProfit,
                remainingCapitalAfter = remainingCapitalAfter,
                monthlyRoi = monthlyRoi
            )
        }
    }

    /**
     * Calculates dynamic financials for a single investment.
     */
    fun calculateInvestmentFinancials(
        investment: Investment,
        records: List<MonthlyRecord>,
        masterCurrency: String,
        rateInfo: CurrencyRateInfo
    ): InvestmentFinancials {
        val initialCapital = investment.initialCapital
        var additionalCapital = 0.0
        var withdrawals = 0.0
        var totalReturn = 0.0
        var totalExpense = 0.0

        records.forEach { rec ->
            additionalCapital += rec.additionalCapital
            withdrawals += rec.withdrawal
            totalReturn += rec.income
            totalExpense += rec.expense
        }

        val totalInvestedCapital = initialCapital + additionalCapital - withdrawals
        val netProfit = totalReturn - totalExpense
        val roi = if (totalInvestedCapital > 0.0) (netProfit / totalInvestedCapital) * 100.0 else 0.0

        val capitalRecovered = totalReturn
        val remainingCapital = max(0.0, totalInvestedCapital - capitalRecovered)

        val recoveryPercentage = if (totalInvestedCapital > 0.0) {
            min(100.0, max(0.0, (capitalRecovered / totalInvestedCapital) * 100.0))
        } else {
            0.0
        }

        var recoveryStatus = CapitalRecoveryStatus.IN_PROGRESS
        if (totalInvestedCapital > 0.0 && capitalRecovered >= totalInvestedCapital) {
            recoveryStatus = if (netProfit > 0.0) {
                CapitalRecoveryStatus.PROFIT_PHASE
            } else {
                CapitalRecoveryStatus.RECOVERED
            }
        }

        var status = InvestmentFinancialStatus.NO_ACTIVITY
        if (totalInvestedCapital == 0.0 && records.isEmpty()) {
            status = InvestmentFinancialStatus.AWAITING_CAPITAL
        } else if (records.isEmpty()) {
            status = InvestmentFinancialStatus.NO_ACTIVITY
        } else if (netProfit > 0.0) {
            status = InvestmentFinancialStatus.PROFIT_GENERATED
        } else if (netProfit < 0.0) {
            status = InvestmentFinancialStatus.OPERATING_AT_LOSS
        } else if (netProfit == 0.0 && (totalReturn > 0.0 || totalExpense > 0.0)) {
            status = InvestmentFinancialStatus.BREAK_EVEN
        }

        // Target period check
        val targetPeriod = investment.targetPeriod
        val startDateStr = investment.startDate
        if (!targetPeriod.isNullOrBlank() && startDateStr.isNotBlank()) {
            try {
                val start = dateFormat.parse(startDateStr)?.time
                if (start != null) {
                    val now = System.currentTimeMillis()
                    val regex = Regex("""(\d+)\s*(month|year)""", RegexOption.IGNORE_CASE)
                    val match = regex.find(targetPeriod)
                    if (match != null) {
                        val num = match.groupValues[1].toIntOrNull() ?: 0
                        val isYear = match.groupValues[2].startsWith("year", ignoreCase = true)
                        val targetMonths = if (isYear) num * 12 else num
                        val targetEndDate = start + (targetMonths * 30.4375 * 24 * 60 * 60 * 1000).toLong()
                        val daysLeft = (targetEndDate - now) / (1000.0 * 60 * 60 * 24)

                        if (daysLeft < 0 && recoveryStatus != CapitalRecoveryStatus.RECOVERED && recoveryStatus != CapitalRecoveryStatus.PROFIT_PHASE) {
                            status = InvestmentFinancialStatus.TARGET_MISSED
                        } else if (daysLeft in 0.0..60.0 && recoveryStatus == CapitalRecoveryStatus.IN_PROGRESS) {
                            status = InvestmentFinancialStatus.TARGET_APPROACHING
                        }
                    }
                }
            } catch (_: Exception) { }
        }

        // Currency conversions
        val rateToMaster = CurrencyHelper.getExchangeRate(investment.currency, masterCurrency, rateInfo.rates)
        val convertedTotalInvested = CurrencyHelper.convertCurrency(totalInvestedCapital, investment.currency, masterCurrency, rateInfo.rates)
        val convertedTotalReturn = CurrencyHelper.convertCurrency(totalReturn, investment.currency, masterCurrency, rateInfo.rates)
        val convertedTotalExpense = CurrencyHelper.convertCurrency(totalExpense, investment.currency, masterCurrency, rateInfo.rates)
        val convertedNetProfit = CurrencyHelper.convertCurrency(netProfit, investment.currency, masterCurrency, rateInfo.rates)
        val convertedCapitalRecovered = CurrencyHelper.convertCurrency(capitalRecovered, investment.currency, masterCurrency, rateInfo.rates)
        val convertedRemainingCapital = CurrencyHelper.convertCurrency(remainingCapital, investment.currency, masterCurrency, rateInfo.rates)

        return InvestmentFinancials(
            initialCapital = initialCapital,
            additionalCapital = additionalCapital,
            withdrawals = withdrawals,
            totalInvestedCapital = totalInvestedCapital,
            totalReturn = totalReturn,
            totalExpense = totalExpense,
            netProfit = netProfit,
            roi = roi,
            capitalRecovered = capitalRecovered,
            remainingCapital = remainingCapital,
            recoveryPercentage = recoveryPercentage,
            recoveryStatus = recoveryStatus,
            status = status,
            rateToMaster = rateToMaster,
            rateDate = rateInfo.date,
            rateSource = rateInfo.source,
            isCachedRate = rateInfo.isFallback,
            convertedTotalInvested = convertedTotalInvested,
            convertedTotalReturn = convertedTotalReturn,
            convertedTotalExpense = convertedTotalExpense,
            convertedNetProfit = convertedNetProfit,
            convertedCapitalRecovered = convertedCapitalRecovered,
            convertedRemainingCapital = convertedRemainingCapital
        )
    }

    /**
     * Consolidates all portfolio assets into Master Currency.
     */
    fun calculatePortfolioSummary(
        investments: List<Investment>,
        recordsByInvestment: Map<String, List<MonthlyRecord>>,
        masterCurrency: String,
        rateInfo: CurrencyRateInfo
    ): PortfolioSummary {
        var totalInvestedCapital = 0.0
        var totalReturn = 0.0
        var totalExpense = 0.0
        var netProfit = 0.0
        var capitalRecovered = 0.0
        var remainingCapital = 0.0

        investments.forEach { inv ->
            val records = recordsByInvestment[inv.id] ?: emptyList()
            val fin = calculateInvestmentFinancials(inv, records, masterCurrency, rateInfo)

            totalInvestedCapital += fin.convertedTotalInvested
            totalReturn += fin.convertedTotalReturn
            totalExpense += fin.convertedTotalExpense
            netProfit += fin.convertedNetProfit
            capitalRecovered += fin.convertedCapitalRecovered
            remainingCapital += fin.convertedRemainingCapital
        }

        val roi = if (totalInvestedCapital > 0.0) (netProfit / totalInvestedCapital) * 100.0 else 0.0
        val recoveryPercentage = if (totalInvestedCapital > 0.0) {
            min(100.0, max(0.0, (capitalRecovered / totalInvestedCapital) * 100.0))
        } else {
            0.0
        }

        var recoveryStatus = CapitalRecoveryStatus.IN_PROGRESS
        if (totalInvestedCapital > 0.0 && capitalRecovered >= totalInvestedCapital) {
            recoveryStatus = if (netProfit > 0.0) {
                CapitalRecoveryStatus.PROFIT_PHASE
            } else {
                CapitalRecoveryStatus.RECOVERED
            }
        }

        return PortfolioSummary(
            masterCurrency = masterCurrency,
            totalInvestedCapital = totalInvestedCapital,
            totalReturn = totalReturn,
            totalExpense = totalExpense,
            netProfit = netProfit,
            roi = roi,
            capitalRecovered = capitalRecovered,
            remainingCapital = remainingCapital,
            recoveryPercentage = recoveryPercentage,
            recoveryStatus = recoveryStatus,
            activeInvestmentsCount = investments.size
        )
    }

    /**
     * Calculates funding summary and source breakdowns.
     */
    fun calculateFundingSummary(
        records: List<FundingRecord>,
        masterCurrency: String,
        rateInfo: CurrencyRateInfo
    ): FundingSummary {
        var totalFunded = 0.0
        var totalAllocated = 0.0
        var totalDistributed = 0.0

        val sourceTotals = mutableMapOf<String, Pair<Double, Int>>()

        records.forEach { rec ->
            val convertedAmount = CurrencyHelper.convertCurrency(rec.amount, rec.currency, masterCurrency, rateInfo.rates)
            val type = rec.type.lowercase()

            if (type.contains("distribution") || type.contains("withdrawal")) {
                totalDistributed += convertedAmount
            } else {
                totalFunded += convertedAmount
                if (!rec.allocatedInvestmentId.isNullOrBlank()) {
                    totalAllocated += convertedAmount
                }

                val current = sourceTotals[rec.source] ?: Pair(0.0, 0)
                sourceTotals[rec.source] = Pair(current.first + convertedAmount, current.second + 1)
            }
        }

        val unallocatedReserve = max(0.0, totalFunded - totalAllocated - totalDistributed)

        val breakdown = sourceTotals.map { (source, pair) ->
            val share = if (totalFunded > 0.0) (pair.first / totalFunded) * 100.0 else 0.0
            SourceBreakdown(
                source = source,
                totalAmountMaster = pair.first,
                count = pair.second,
                sharePercent = share
            )
        }.sortedByDescending { it.totalAmountMaster }

        return FundingSummary(
            totalFundedMaster = totalFunded,
            totalAllocatedMaster = totalAllocated,
            unallocatedReserveMaster = unallocatedReserve,
            totalDistributedMaster = totalDistributed,
            sourceBreakdown = breakdown
        )
    }
}

File: src/lib/calculations.ts
import {
  Investment,
  MonthlyRecord,
  ComputedMonthlyRecord,
  InvestmentFinancials,
  PortfolioSummary,
  CapitalRecoveryStatus,
  InvestmentFinancialStatus,
  CurrencyRateInfo,
} from '../types';
import { getExchangeRate, convertCurrency } from './currency';

/**
 * Computes spreadsheet-style running monthly values for an investment.
 */
export function computeMonthlyRecords(
  investment: Investment,
  records: MonthlyRecord[]
): ComputedMonthlyRecord[] {
  // Sort records chronologically by date
  const sorted = [...records].sort((a, b) => new Date(a.date).getTime() - new Date(b.date).getTime());

  let cumulativeCapital = investment.initialCapital;
  let cumulativeRecovered = 0;

  return sorted.map((rec) => {
    const netProfit = (rec.income || 0) - (rec.expense || 0);
    cumulativeCapital += (rec.additionalCapital || 0) - (rec.withdrawal || 0);
    cumulativeRecovered += (rec.income || 0);

    const remainingCapitalAfter = Math.max(0, cumulativeCapital - cumulativeRecovered);
    const monthlyRoi = cumulativeCapital > 0 ? (netProfit / cumulativeCapital) * 100 : 0;

    return {
      ...rec,
      netProfit,
      remainingCapitalAfter,
      monthlyRoi,
    };
  });
}

/**
 * Calculates comprehensive dynamic financials for a single investment.
 * First in its native currency, then converted to master currency.
 */
export function calculateInvestmentFinancials(
  investment: Investment,
  records: MonthlyRecord[],
  masterCurrency: string,
  rateInfo: CurrencyRateInfo
): InvestmentFinancials {
  const initialCapital = Number(investment.initialCapital) || 0;
  
  let additionalCapital = 0;
  let withdrawals = 0;
  let totalReturn = 0;
  let totalExpense = 0;

  records.forEach((rec) => {
    additionalCapital += Number(rec.additionalCapital) || 0;
    withdrawals += Number(rec.withdrawal) || 0;
    totalReturn += Number(rec.income) || 0;
    totalExpense += Number(rec.expense) || 0;
  });

  const totalInvestedCapital = initialCapital + additionalCapital - withdrawals;
  const netProfit = totalReturn - totalExpense;
  const roi = totalInvestedCapital > 0 ? (netProfit / totalInvestedCapital) * 100 : 0;

  // Capital Recovered is the total returns received back
  const capitalRecovered = totalReturn;
  const remainingCapital = Math.max(0, totalInvestedCapital - capitalRecovered);
  
  const recoveryPercentage = totalInvestedCapital > 0 
    ? Math.min(100, Math.max(0, (capitalRecovered / totalInvestedCapital) * 100))
    : 0;

  // Capital Recovery Status
  let recoveryStatus: CapitalRecoveryStatus = 'Capital Recovery in Progress';
  if (totalInvestedCapital > 0 && capitalRecovered >= totalInvestedCapital) {
    if (netProfit > 0) {
      recoveryStatus = 'Profit Phase';
    } else {
      recoveryStatus = 'Capital Recovered';
    }
  }

  // Investment Financial Status
  let status: InvestmentFinancialStatus = 'No Activity';
  if (totalInvestedCapital === 0 && records.length === 0) {
    status = 'Awaiting Capital';
  } else if (records.length === 0) {
    status = 'No Activity';
  } else if (netProfit > 0) {
    status = 'Profit Generated';
  } else if (netProfit < 0) {
    status = 'Operating at Loss';
  } else if (netProfit === 0 && (totalReturn > 0 || totalExpense > 0)) {
    status = 'Break-even';
  }

  // Check Target Period if provided
  if (investment.targetPeriod && investment.startDate) {
    const start = new Date(investment.startDate).getTime();
    const now = Date.now();
    const monthsMatch = investment.targetPeriod.match(/(\d+)\s*(month|year)/i);
    if (monthsMatch) {
      const num = parseInt(monthsMatch[1], 10);
      const isYear = monthsMatch[2].toLowerCase().startsWith('year');
      const targetMonths = isYear ? num * 12 : num;
      const targetEndDate = start + targetMonths * 30.4375 * 24 * 60 * 60 * 1000;
      const daysLeft = (targetEndDate - now) / (1000 * 60 * 60 * 24);

      if (daysLeft < 0 && recoveryStatus !== 'Capital Recovered' && recoveryStatus !== 'Profit Phase') {
        status = 'Target Period Missed';
      } else if (daysLeft >= 0 && daysLeft <= 60 && recoveryStatus === 'Capital Recovery in Progress') {
        status = 'Target Period Approaching';
      }
    }
  }

  // Currency Conversion to Master Currency
  const rateToMaster = getExchangeRate(investment.currency, masterCurrency, rateInfo.rates);
  const convertedTotalInvested = convertCurrency(totalInvestedCapital, investment.currency, masterCurrency, rateInfo.rates);
  const convertedTotalReturn = convertCurrency(totalReturn, investment.currency, masterCurrency, rateInfo.rates);
  const convertedTotalExpense = convertCurrency(totalExpense, investment.currency, masterCurrency, rateInfo.rates);
  const convertedNetProfit = convertCurrency(netProfit, investment.currency, masterCurrency, rateInfo.rates);
  const convertedCapitalRecovered = convertCurrency(capitalRecovered, investment.currency, masterCurrency, rateInfo.rates);
  const convertedRemainingCapital = convertCurrency(remainingCapital, investment.currency, masterCurrency, rateInfo.rates);

  return {
    initialCapital,
    additionalCapital,
    withdrawals,
    totalInvestedCapital,
    totalReturn,
    totalExpense,
    netProfit,
    roi,
    capitalRecovered,
    remainingCapital,
    recoveryPercentage,
    recoveryStatus,
    status,
    rateToMaster,
    rateDate: rateInfo.date,
    rateSource: rateInfo.source,
    isCachedRate: !!rateInfo.isFallback,
    convertedTotalInvested,
    convertedTotalReturn,
    convertedTotalExpense,
    convertedNetProfit,
    convertedCapitalRecovered,
    convertedRemainingCapital,
  };
}

/**
 * Consolidates all investments in the portfolio into the user's Master Currency.
 */
export function calculatePortfolioSummary(
  investments: Investment[],
  recordsByInvestment: Record<string, MonthlyRecord[]>,
  masterCurrency: string,
  rateInfo: CurrencyRateInfo
): PortfolioSummary {
  let totalInvestedCapital = 0;
  let totalReturn = 0;
  let totalExpense = 0;
  let netProfit = 0;
  let capitalRecovered = 0;
  let remainingCapital = 0;

  investments.forEach((inv) => {
    const records = recordsByInvestment[inv.id] || [];
    const financials = calculateInvestmentFinancials(inv, records, masterCurrency, rateInfo);
    
    totalInvestedCapital += financials.convertedTotalInvested;
    totalReturn += financials.convertedTotalReturn;
    totalExpense += financials.convertedTotalExpense;
    netProfit += financials.convertedNetProfit;
    capitalRecovered += financials.convertedCapitalRecovered;
    remainingCapital += financials.convertedRemainingCapital;
  });

  const roi = totalInvestedCapital > 0 ? (netProfit / totalInvestedCapital) * 100 : 0;
  const recoveryPercentage = totalInvestedCapital > 0 
    ? Math.min(100, Math.max(0, (capitalRecovered / totalInvestedCapital) * 100))
    : 0;

  let recoveryStatus: CapitalRecoveryStatus = 'Capital Recovery in Progress';
  if (totalInvestedCapital > 0 && capitalRecovered >= totalInvestedCapital) {
    if (netProfit > 0) {
      recoveryStatus = 'Profit Phase';
    } else {
      recoveryStatus = 'Capital Recovered';
    }
  }

  return {
    masterCurrency,
    totalInvestedCapital,
    totalReturn,
    totalExpense,
    netProfit,
    roi,
    capitalRecovered,
    remainingCapital,
    recoveryPercentage,
    recoveryStatus,
    activeInvestmentsCount: investments.length,
  };
}
